#162948846 Users should be able to follow each other

andrewhingah opened the pull request:
**What does this PR do?**
Creates _friends_ application to handle following and unfollowing users

**Description of tasks to be completed**

- Create a Friend model class to handle user relationships
- Create a CustomSerializer to verify the usernames
- Write the views
- Add the routes

**Any background context you want to provide?**

- A user should be able to follow and be followed by other users
- A user can unfollow and be unfollowed by other users
- A user can view his/her followers (or other users followers)
- A user can see who they are following (or who other users are following)

**How should this be manually tested?**

- `git clone https://github.com/andela/ah-django-unchained`
- `cd ah-django-unchained`
- `git checkkout ft-user-follow-unfollow-162948846`
- `python3 -m venv virtual_env_name` to create a virtual environment
- Add a `.env` file as indicated in the ReadMe
- `source .env` to load your environment variables
- `pip install -r requirements.txt`
- Run the migrations `python manage.py makemigrations && python manage.py migrate`

On Postman, register 2 or more users via the endpoint `http://127.0.0.1:8000/api/users/`.

To follow a particular user
- Make a `POST` request on `http://127.0.0.1:8000/api/profiles/<username>/follow` where the username should be of the user you intend to follow, ensure to copy the token of one of the other user and add it as an authorization header in the request. (In all the following screenshots the current signed-in user is `Caren`)
- The response returns the user id of the user being followed.
<img width="970" alt="screenshot 2019-01-18 at 15 20 39" src="https://user-images.githubusercontent.com/32644295/51386794-c6dafb00-1b34-11e9-8f2c-ff59e662b5c3.png">



To view the followers of the user that you just followed,
- Make a `GET` request on `http://127.0.0.1:8000/api/profiles/<username>/followers` where the username should be of the user you just followed.

<img width="968" alt="screenshot 2019-01-18 at 15 22 44" src="https://user-images.githubusercontent.com/32644295/51386865-01dd2e80-1b35-11e9-84c7-807271f15440.png">



To view who you are following,
- Make a `GET` request on `http://127.0.0.1:8000/api/profiles/<username>/following` where the username should be of the current signed in user.
<img width="962" alt="screenshot 2019-01-18 at 15 24 16" src="https://user-images.githubusercontent.com/32644295/51386913-29cc9200-1b35-11e9-9a69-3c672e483019.png">


To unfollow a particular user
- Use url `http://127.0.0.1:8000/api/profiles/<username>/follow` with `DELETE` where the username should be of the user you followed, make sure you copy the token from the registration response and add it as an authorization header in the request
- The response returns the user id of the user being unfollowed.

<img width="968" alt="screenshot 2019-01-18 at 15 16 56" src="https://user-images.githubusercontent.com/32644295/51386654-40beb480-1b34-11e9-8450-4f0a3b9f17fe.png">



**What are the relevant Pivotal Tracker Stories?**
[#162948846](https://www.pivotaltracker.com/story/show/162948846)

andrewhingah committed:
feat(User following): Add unit tests for user following

- Create app for users to be able to follow each other
- Write unit tests for this functionality

[Delivers #162948846]

andrewhingah committed:
feat(User Following): implement follow fuctionality

- create endpoint for user to follow another user
- create endpoint for users to view followers and who they're following
- prevent self following
- create an endpoint to  unfollow

[Delivers #162948846]

andrewhingah committed:
feat(user following): refactor tests

- remove errors in tests
- put authenticated user and new user in setup

[Delivers #162948846]

andrewhingah committed:
feat(user following): register Friend model in admin

- register Friend model in admin
- correct spelling mistakes

[Fixes #162948846]

andrewhingah committed:
feat(user following) add more tests

- test for unfollow self
- test unfollow non existing user

[Delivers #162948846]

andrewhingah committed:
feat(user following): return user being followed/unfollowed id

- add id field in user serializer
- modify response in views to return user id

[Fixes #162948846]

kwameasiago reviewed (approved):
All is Good.

ignatiusukwuoma commented on authors/apps/authentication/serializers.py:
```diff
@@ -144,6 +150,11 @@ class Meta:
         # `max_length` properties too, but that isn't the case for the token
         # field.
+    def get_followers(self, obj):
+        return len(json.loads(json.dumps([u.username for u in obj.followers.all()])))
```
Function name should be something like `get_number_of _followers`. Also, for getting the number can we user `count` instead?

andrewhingah replied:
Let me work on that

ignatiusukwuoma replied:
There must be a better way to do this

ignatiusukwuoma commented on authors/apps/authentication/serializers.py:
```diff
@@ -144,6 +150,11 @@ class Meta:
+        return len(json.loads(json.dumps([u.username for u in obj.followers.all()])))
+
+    def get_following(self, obj):
+        return len(json.loads(json.dumps([u.username for u in obj.following.all()])))
```
ditto

ignatiusukwuoma commented on authors/apps/friends/urls.py:
```diff
@@ -0,0 +1,22 @@
+        name='follow'),
+    path(
+        'users/<username>/unfollow',
+        views.UnFollowApiView.as_view(),
```
use one class based view for both follow and unfollow since the README suggest that both actions should use same endpoint

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -0,0 +1,127 @@
+    """view for unfollowing a user"""
+    permission_classes = (IsAuthenticated,)
+
+    def delete(self, request, username, format=None):
```
Have your delete method in the FollowApiView so that both actions use same endpoint

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -0,0 +1,127 @@
+        If it's the same user, then an error message is returned
+        Else the intermediary Friend model will be used to
+        create the user relationship
+        Finally the serialized object of the followed user is returned
```
Nice comments but try to summarise it and make it sound normal

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -0,0 +1,127 @@
+        If it's the same user, then an error message is returned
+        Else the intermediary Friend model will be used to
+        delete the user relationship
+        Finally the serialized object of the unfollowed user is returned
```
ditto

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -0,0 +1,127 @@
+    def get(self, request, username, format=None):
+        """
+            Get & return the followers from the respective
+            object relationships i.e 'user_from'
```
Be consistent in docstrings indentation

andrewhingah committed:
feat(user following): put follow and unfollow in one class based view

- put follow and unfollow in one class based view

[Delivers #162948846]

andrewhingah committed:
feat(user following): docstrings consistency

- fix unconsistency in docstrings indentation
- summarize comments in follow view

[Fixes #162948846]

andrewhingah committed:
feat(user following): rename function names to be descriptive

- rename function names to be descriptive

[Fixes #162948846]

ignatiusukwuoma commented on authors/apps/friends/models.py:
```diff
@@ -0,0 +1,37 @@
+            self.user_to.username)
+
+# adds following field to user dynamically
+get_user_model().add_to_class('following', models.ManyToManyField(
```
Do we need a followers field on User since we already have Friends model that keeps the relationship?

andrewhingah replied:
Sorry this is meant for displaying the relationships in the admin page. I can add that to the PR description.

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -0,0 +1,120 @@
+            return Response(
+                {"message": "You cannot follow yourself"},
+                status=status.HTTP_406_NOT_ACCEPTABLE)
+        Friend.objects.get_or_create(user_from=follower, user_to=followed)
```
What happens if i try to follow a user that i have followed before?

andrewhingah replied:
For now, no error is returned but the followers count of that user remains the same. What would be the best way to handle that?

andrewhingah replied:
I have now worked on this. A user will get a message saying they already follow the user

ignatiusukwuoma replied:
Since you are using get_or_create, if it gets the value and `created` returns False, then return a Response that user has already being followed. See https://docs.djangoproject.com/en/2.1/ref/models/querysets/#get-or-create
In that case, you return early and do not have to execute lines 37 - 41

andrewhingah committed:
feat(user following): feedback implementation

- create custom serializer to return number of followers or following
- test number of following and number of followers after
- a user is followed or unfollowed

[Fixes #162948846]

ignatiusukwuoma commented on authors/apps/friends/serializers.py:
```diff
@@ -0,0 +1,35 @@
+            )
+
+    def get_number_of_followers(self, obj):
+        return len(
```
There's a much efficient way to do this... find it.

andrewhingah replied:
Could you recommend a better way?

ignatiusukwuoma replied:
I thought I did. Didn't you understand this?
https://docs.djangoproject.com/en/2.1/ref/models/querysets/#when-querysets-are-evaluated

andrewhingah replied:
George has assisted me with this. See the changes.

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -0,0 +1,120 @@
+                status=status.HTTP_406_NOT_ACCEPTABLE)
+        Friend.objects.get_or_create(user_from=follower, user_to=followed)
+        user = get_user_model().objects.get(pk=followed.id)
+        serilizer = FollowUnfollowSerializer(user)
```
Can you correctly spell what is intended? `serializer` instead of `serilizer`
 Doesn't matter much but it feels better

andrewhingah committed:
feat(user following): add error messages

- add error msg for user cannot follow a user they already follow
- add error msg for user cannot unfollow a user they dont follow
- add tests for the error messages

[Delivers #162948846]

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -0,0 +1,120 @@
+        user = get_user_model().objects.get(pk=followed.id)
+        serilizer = FollowUnfollowSerializer(user)
+        user_id = serilizer.data
+        if serilizer:
```
Lines 39 and 40 doesn't look necessary to be just return serializer.data as response

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -0,0 +1,135 @@
+                {"message": "You cannot follow yourself"},
+                status=status.HTTP_406_NOT_ACCEPTABLE)
+
+        if Friend.objects.filter(user_from=follower, user_to=followed):
```
See my comment on `get or create` below

ignatiusukwuoma replied:
This line is not required. get_or_create returns a tuple that tells you if that relationship exists previously. If it does, return this response otherwise continue with the rest of the code.

andrewhingah replied:
I added a condition to check if the boolean in the tuple is `False` to notify one has already followed the user.

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -0,0 +1,135 @@
+        """
+        friend_objects = self.get_queryset()
+        if friend_objects is not None:
+            users_followed = {u.user_to for u in friend_objects}
```
List comprehension is to create lists right? users_followed should be a list not dictionary

andrewhingah replied:
Fixing that right away.

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -0,0 +1,135 @@
+        user = get_object_or_404(
+            get_user_model(), username=self.kwargs['username'])
+        return Friend.objects.select_related(
+            'user_from', 'user_to', ).filter(user_from=user.id).all()
```
I don't think we need `'user_to'` here

andrewhingah replied:
Thanks. Let me fix that

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -0,0 +1,135 @@
+                status=status.HTTP_406_NOT_ACCEPTABLE)
+
+        Friend.objects.get_or_create(user_from=follower, user_to=followed)
+        user = get_user_model().objects.get(pk=followed.id)
```
Is the value of `user` not the same as `followed`? Do we need to hit the database again?

andrewhingah replied:
Oh yes they are the same value. No need to hit the db again.

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -0,0 +1,135 @@
+        user = get_user_model().objects.get(pk=followed.id)
+        serializer = FollowUnfollowSerializer(user)
+        user_id = serializer.data
+        if serializer:
```
Is there a scenario where serializer will be a falsy value? I feel code is more readable without line 80 and 81.

andrewhingah replied:
I just realised that is not necessary. Let me get rid of it.

andrewhingah committed:
feat(user following): Use count method to get number of followers and following

- use count to get the number of users
- remove unnecessary condition in serializer
- fix list comprehesion in user_followed

[Delivers #162948846]

ignatiusukwuoma commented on authors/apps/friends/serializers.py:
```diff
@@ -24,12 +24,8 @@ class Meta:
-            json.loads(json.dumps([u.username for u in obj.followers.all()]))
-            )
+    def get_number_of_followers(self, followers):
+        return followers
```
Hmm.. your serializer doesn't look like it's doing anything now, just returning what was passed to it.

andrewhingah replied:
Working on it.

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -14,6 +14,16 @@ class FollowUnfollowApiView(generics.RetrieveUpdateDestroyAPIView):
     """view for following a user"""
     permission_classes = (IsAuthenticated,)
 
+    def get_number_of_followers(self, user):
```
Let's follow the Single Responsibility principle and handle `get_number_of_followers` in one place. The only change I expected you to make is to refactor the function on the serializer and not add a new function to the view. This function would have worked fine but you don't need `select-related` since you're only looking for a count. Read more on select-related here https://docs.djangoproject.com/en/2.1/ref/models/querysets/#select-related.
No need for these changes, just returning `obj.followers.count()` is all you need.

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -34,17 +44,23 @@ def post(self, request, username, format=None):
                 status=status.HTTP_406_NOT_ACCEPTABLE)
 
-        if Friend.objects.filter(user_from=follower, user_to=followed):
+        is_following = Friend.objects.get_or_create(
```
While this works, it's more readable to assign different variables to the return value of a function that returns a tuple. Something like 
`user_followed, created_now = Friend.objects.get_or_create...`

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -34,17 +44,23 @@ def post(self, request, username, format=None):
-        if Friend.objects.filter(user_from=follower, user_to=followed):
+        is_following = Friend.objects.get_or_create(
+            user_from=follower, user_to=followed)
+        if is_following[1] is False:
```
If you followed the above, you can use `created_now` here instead

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -34,17 +44,23 @@ def post(self, request, username, format=None):
-        user_id = serializer.data
-        if serializer:
-            return Response(user_id, status=status.HTTP_200_OK)
+        number_of_followers = self.get_number_of_followers(followed)
```
I don't think we need to include these in the response message. It feels like unnecessary db hits. I'm fine with the previous implementation asides the comment i made.

andrewhingah committed:
feat(user following): handle get number of followes in one place

- use obj.<name>.count() method to return number of users

[Fixes #162948846]

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -0,0 +1,132 @@
+        user = get_object_or_404(
+            get_user_model(), username=self.kwargs['username'])
+        return Friend.objects.select_related(
+            'user_to').filter(user_to=user.id).all()
```
The select_related field here should be `user_from` because that's what you use in line 102

ignatiusukwuoma commented on authors/apps/friends/views.py:
```diff
@@ -0,0 +1,132 @@
+        user = get_object_or_404(
+            get_user_model(), username=self.kwargs['username'])
+        return Friend.objects.select_related(
+            'user_from', ).filter(user_from=user.id).all()
```
The select_related field here should be `user_to` because that's what you use in line 127
